### ISSUE TITLE
[RFC] Sleep-mode tensor ownership and recovery on Ascend

### ISSUE THREAD

**Issue by Ronald1995** [vllm-project/vllm-ascend#16558]
# RFC: Sleep-mode tensor ownership and recovery on Ascend

## Summary

vLLM Ascend sleep mode currently assigns an allocation tag according to the
active `CaMemAllocator` pool.  That tag controls whether an allocation is
host-backed, discarded, or kept mapped across sleep/wake.  It does not describe
the semantic lifetime of every tensor that shares the allocation scope.

This RFC proposes a semantic ownership contract for persistent NPU tensors and
a narrower `kv_cache` allocation boundary.  Its primary objective is to ensure
that a tensor that is needed after wake-up is never silently discarded merely
because it was created while the KV-cache pool was active.

This is the Ascend counterpart of the upstream vLLM RFC
[Sleep Mode Tensor Ownership and Recovery](https://github.com/vllm-project/vllm/issues/53343).
It builds on the targeted fixes already merged for DSA Hadamard matrices and
the ALLTOALL expert-ID map in vLLM Ascend PR #13352.

## Motivation

`CaMemAllocator` records one tag per underlying allocation.  In the current
worker path, `Worker.initialize_from_config()` enters
`use_memory_pool(tag="kv_cache")` around the complete
`model_runner.initialize_kv_cache()` call.  This operation does more than
allocate cache backing storage.  Depending on the runner and configuration, it
also creates or binds attention backends, input-batch state, speculative
decoding state, KV-transfer registrations, sparse-KV state, KVPP runtime state,
and ACL-graph-related metadata.

At level 1, allocations tagged `weights` are copied to host memory; other
non-persistent allocations are unmapped and released.  At level 2, weights are
also discarded and are expected to be reloaded.  `sleep_persistent` allocations
remain mapped.  On wake-up, `CaMemAllocator` recreates storage at the original
virtual address, but only host-backed allocations have their contents copied
back.  Thus pointer stability does not imply value preservation.

This is especially hazardous for ACL graphs: a replay can retain valid device
pointers while reading an erased scale, index, sentinel, HCCL control value, or
device pointer array.  The result can be silent accuracy loss, an incorrect
routing decision, or a hang rather than an invalid-pointer failure.

The repository already has useful mitigations:

- level-2 wake-up saves and restores registered model buffers;
- `sleep_persistent` prevents selected runtime allocations from being released;
- `get_or_register_attention_buffer()` attaches shared attention tensors to
  their owning modules as non-persistent buffers;
- KV-zero bookkeeping is intentionally initialized outside the current
  `kv_cache` pool; and
- optional extra cleanup invalidates and recaptures ACL graphs and rebuilds
  HCCL state.

These mechanisms repair individual cases but do not yet establish an allocation
boundary or a universal recovery contract.

## Goals

1. Prevent a persistent non-KV tensor from being silently discarded by either
   sleep level.
2. Make recovery explicit for model buffers, derived weights, runtime metadata,
   communication state, and stateful workspaces.
3. Preserve device-address stability required by ACL graphs, or explicitly
   invalidate and recapture graphs when stability cannot be preserved.
4. Continue reclaiming actual KV-cache and request-dependent state.
5. Add deterministic tests that reveal missing recovery even when fresh NPU
   memory happens to be zero-filled.
6. Provide the same contract to Ascend attention backends and KV-transfer
   connectors.

## Non-goals

- Preserve active-request KV contents across sleep.
- Automatically reload level-2 model parameters; the existing caller-driven
  reload/update sequence remains unchanged.
- Preserve HCCL or ACL graph resources when extra cleanup explicitly destroys
  them.
- Force every scratch allocation to be initialized.  A scratch owner may discard
  its storage only when it completely writes it before every read.

## Tensor ownership contract

Every NPU tensor that outlives initialization must have one semantic owner and
one of the following recovery classes.

| Class | Sleep behavior | Required recovery |
| --- | --- | --- |
| KV/recurrent cache backing | Discardable | No new request may read pre-sleep data. |
| Model parameter | Level 1 host-backed; level 2 discarded | Caller reloads it before execution after level 2. |
| Static model tensor | Recoverable | Register it as a model buffer or rematerialize from host/Python data. |
| Derived weight | Recoverable after reload | Rebuild in a deterministic post-reload hook. |
| Persistent runtime metadata | Remains resident or is restored in place | Allocate outside the discardable pool, use a buffer, or use a narrowly scoped wake hook. |
| Scratch workspace | Discardable | Fully overwrite/reset before its first post-wake read. |
| Communication/synchronization state | Recoverable | Rebuild protocol values, counters, layouts, and pointer arrays before use. |

Required invariants:

1. A tensor in a discardable allocation is itself discardable, or is completely
   initialized before its first post-wake read.
2. Tensors captured by an ACL graph are restored in place.  Replacing the
   Python tensor object requires invalidating and recapturing every affected
   graph.
3. A level-2 parameter reload is followed by rebuilding all tensors that pack,
   transpose, concatenate, or otherwise snapshot those parameters.
4. Synchronization state is restored to its protocol-specific initial state;
   zeroing is not a substitute for restoring sentinels or pointer tables.
5. Connectors registering KV-memory regions may retain only cache backing in
   the discardable pool.  Their control metadata follows this contract.

### 1. Narrow the `kv_cache` allocation scope

Split KV initialization into preparation, cache allocation, and binding:

```python
prepared = model_runner.prepare_kv_cache(kv_cache_config)
with allocator.use_memory_pool(tag="kv_cache"):
    kv_cache_tensors = model_runner.allocate_kv_cache_tensors(prepared)
model_runner.bind_kv_cache(prepared, kv_cache_tensors)
```

Only raw cache backing allocations and their shaped views belong in
`allocate_kv_cache_tensors`.  The following must be prepared or bound outside
the discardable pool unless explicitly classified as scratch: attention backend
metadata, input-batch bookkeeping, speculative state, connector objects and
registration metadata, KVPP runtime state, graph-manager metadata, and fixed
index/scale tensors.

The exact APIs may differ for MRV1, MRV2, and 310P.  Each implementation must
enforce the same boundary rather than merely moving a few known allocations.

### 2. Standardize recovery hooks

Define a small runner-owned lifecycle protocol for state that must be rebuilt
after an allocation has been discarded:

```python
class SleepAwareTensorOwner(Protocol):
    def post_sleep_wake(self, discarded_tags: set[str]) -> None: ...
    def post_weights_reload(self) -> None: ...
```

The worker/model runner keeps an explicit registry of owners.  It must not
traverse arbitrary Python objects to discover tensors.  The ordered lifecycle
is:

1. restore allocator mappings and host-backed bytes;
2. restore registered model buffers;
3. reload parameters when level 2 requires it;
4. rebuild derived weights;
5. reset runtime, HCCL, and synchronization state;
6. recapture ACL graphs when the state or addresses changed; and
7. admit new scheduling/execution.

Existing `post_kv_cache_wake_up()` compatibility behavior should be folded into
this common lifecycle as upstream compatibility allows.

### 3. Prefer model buffers or host values for static state

Device tensors that are static model state must be registered on an owning
`nn.Module` with `persistent=False` when they should not enter checkpoints.
The worker's existing `model.named_buffers()` backup path then covers level 2.
Scalar constants and small ID lists should remain Python/host values where a
device pointer is unnecessary.  Non-module helpers must use their owning model
or layer as the registered-buffer owner, as done by the existing attention
buffer helper and the MoE routed-experts fix.

### 4. Reserve `sleep_persistent` for exceptional runtime state

`sleep_persistent` is a correctness escape hatch for small state that must keep
its mapping and cannot be re-created safely.  It must have a documented owner,
size, and rationale.  It is not a substitute for fixing broad pool scopes,
because it reduces memory reclaimed during sleep and is easy to overuse.

### 5. Make extra-cleanup recovery explicit

When `rl_config.sleep_mode_extra_cleanup` destroys HCCL groups or invalidates
ACL graphs, owner recovery runs before graph capture.  Re-created HCCL metadata
and device pointer arrays must use the newly valid resources.  A component that
cannot update captured state in place must declare graph invalidation.

## Risk inventory and audit plan

The first implementation phase will inventory tensors created while the
`kv_cache` pool is active in MRV1, MRV2, and 310P.  The audit must cover at
least:

- attention backends, including DSA/SFA/MLA and shared attention buffers;
- `NPUInputBatch`, block tables, and KV-zero metadata;
- speculative decoders and Mamba/GDN state;
- sparse-KV-offload and KVPP runtime state;
- KV-transfer connectors and their registration/control buffers;
- MoE ALLTOALL/HCCL state, including expert-ID maps and device pointer arrays;
- ACL graph manager state; and
- plain device-tensor attributes in Ascend model/ops helpers.

For every retained tensor, record its owner, allocation tag, recovery class,
first post-wake reader, graph-capture status, and test coverage.  The output is
the acceptance checklist for the migration, not just a one-time investigation.

## Test strategy

1. **Allocator contract test.**  Allocate a host-backed weight, a KV tensor,
   and unrelated metadata in a `kv_cache` scope.  Sleep/wake and assert stable
   addresses, restored weights, and lost KV/metadata contents.  This documents
   the allocator's allocation-level behavior.
2. **Deterministic poisoning helper.**  In a fresh subprocess, wrap the NPU
   remap operation and fill every newly mapped allocation with a recognizable
   nonzero byte pattern before normal host backup restoration.  Do not rely on
   the observed contents of fresh NPU memory.
3. **ACL-graph reproducer.**  Capture a graph using a nonzero persistent scale
   created in the KV scope.  Verify stable pointers but incorrect output after
   poisoned wake; then verify the corrected boundary or owner recovery produces
   the reference output.
4. **Level-2 classification test.**  Use a toy module with a parameter, a
   non-persistent registered buffer, and a plain tensor attribute.  After
   poisoned level-2 wake and simulated reload, only the parameter and buffer
   may recover.  The permanent regression case converts the plain attribute to
   an owned buffer or explicit recovery owner.
5. **Ascend backend regressions.**  Add compact tests for at least one affected
   attention path, one MoE/HCCL state path, and one connector path.  Check
   contents or model output after two sleep/wake cycles, not only allocation
   validity.
6. **End-to-end sleep/wake.**  Extend the NPU RL sleep/wake suite with a
   deterministic generation or logits comparison before and after sleep for
   configurations exercising the selected backends.  Test both normal sleep and
   extra cleanup with ACL graphs enabled where supported.

### Phase 0: establish failure visibility

- Land the remap-poison helper, allocator contract test, and level-2 ownership
  classification test.
- Add one ACL-graph or backend reproducer that fails under the current broad
  scope.
- Publish the inventory/checklist described above.

### Phase 1: correct allocation boundaries

- Refactor MRV1 so only `_allocate_kv_cache_tensors` runs in `kv_cache`.
- Apply the same preparation/allocation/binding boundary to MRV2 and evaluate
  310P independently, preserving unsupported-feature guards.
- Move current post-pool bookkeeping, connector registration, and runtime
  construction into the binding phase.
- Add an allocation audit asserting that only cache backing is tagged
  `kv_cache` in representative configurations.

### Phase 2: migrate persistent owners

- Convert confirmed static tensors to registered buffers or host values.
- Create explicit owner hooks for derived weights, DSA/SFA metadata, and
  communication state that cannot leave the pool.
- Document each required `sleep_persistent` allocation and remove it when a
  normal owner/recovery strategy becomes viable.
- Make graph invalidation/recapture part of every extra-cleanup owner's test.

### Phase 3: enforce and maintain the contract

- Add a debug-only allocation audit that records tag, allocation range, size,
  creation stack, and weak tensor references during tagged scopes.
- Report surviving tensors at scope exit to backend and connector authors.
- Require the ownership checklist and an appropriate poison regression for new
  persistent NPU state introduced by sleep-mode-capable components.

## Alternatives considered

**Back up all `kv_cache` allocations.**  This preserves accidental metadata but
also copies the complete KV cache to host memory, negating level-1 sleep's
memory and latency benefit.

**Keep every uncertain tensor under `sleep_persistent`.**  This is a useful
temporary mitigation but leaks the semantic error into memory accounting and
reduces reclaimed NPU memory indefinitely.

**Zero newly mapped allocations.**  This gives deterministic behavior but does
not restore scales, indices, sentinels, or device pointers, and can hide bugs
in zero-initialized workspaces.

**Only add component-specific wake hooks.**  Hooks are necessary for protocol
state, but without a narrow allocation boundary every new backend or connector
can repeat the same bug.

**Always recapture ACL graphs.**  This avoids some stale-state cases but
increases wake latency and does not restore eager-mode state.  Recapture is
required only when in-place recovery is impossible or extra cleanup destroyed
the graph resources.

## Compatibility and performance criteria

The migration must preserve the public sleep API and its two-phase level-2
wake-up sequence.  Before/after measurements must report peak model-load NPU
memory, available KV-cache memory, memory freed at each sleep level, retained
`sleep_persistent` bytes, and wake latency with and without extra cleanup.

The expected trade-off is a small amount of retained metadata in exchange for
correctness.  If a backend's recoverable metadata is material, introduce a
separate recoverable policy only after it has a documented owner and a measured
benefit; do not reuse the discardable `kv_cache` tag.

## Acceptance criteria

- No persistent non-KV tensor is allocated in the discardable KV-cache scope
  without a documented complete-rewrite guarantee.
- All identified persistent NPU tensors have an owner and recovery class.
- Poisoned level-1 and level-2 tests pass for the selected representative
  attention, MoE/HCCL, connector, and ACL-graph paths.
- ACL graph tests prove either in-place restoration or explicit invalidation
  and recapture.
- Memory-reclamation and wake-latency measurements are included in each phase's
  pull request series.
